src/sinks/greptimedb/client.rs

```diff
@@ -0,0 +1,533 @@
+use crate::sinks::util::statistic::DistributionStatistic;
+use crate::sinks::util::{EncodedEvent, TowerRequestConfig};
+use crate::sinks::VectorSink;
+
+#[derive(Debug)]
+pub struct GreptimeBatchOutput(Vec<u32>);
+
+impl Response for GreptimeBatchOutput {}
+
+#[derive(Clone)]
+struct GreptimeDBRetryLogic;
+
+impl RetryLogic for GreptimeDBRetryLogic {
+    type Error = GreptimeError;
+    type Response = GreptimeBatchOutput;
+
+    fn is_retriable_error(&self, _error: &Self::Error) -> bool {
+        true
+    }
+}
+
+#[derive(Clone, Debug, Default)]
+struct GreptimeDBMetricNormalize;
+
+impl MetricNormalize for GreptimeDBMetricNormalize {
+    fn normalize(&mut self, state: &mut MetricSet, metric: Metric) -> Option<Metric> {
+        match (metric.kind(), &metric.value()) {
+            (_, MetricValue::Counter { .. }) => state.make_absolute(metric),
+            (_, MetricValue::Gauge { .. }) => state.make_absolute(metric),
+            // All others are left as-is
+            _ => Some(metric),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct GreptimeDBService {
+    /// the client that connects to greptimedb
+    client: Arc<Database>,
+}
+
+impl GreptimeDBService {
+    pub fn new_sink(config: &GreptimeDBConfig) -> crate::Result<VectorSink> {
+        let grpc_client = Client::with_urls(vec![&config.endpoint]);
+        let mut client = Database::new_with_dbname(
+            config.dbname.as_deref().unwrap_or(DEFAULT_SCHEMA_NAME),
+            grpc_client,
+        );
+
+        if let (Some(username), Some(password)) = (&config.username, &config.password) {
+            client.set_auth(AuthScheme::Basic(Basic {
+                username: username.to_owned(),
+                password: password.clone().into(),
+            }))
+        }
+
+        let batch = config.batch.into_batch_settings()?;
+        let request = config.request.unwrap_with(&TowerRequestConfig {
+            retry_attempts: Some(1),
+            ..Default::default()
+        });
+
+        let greptime_service = GreptimeDBService {
+            client: Arc::new(client),
+        };
+
+        let mut normalizer = MetricNormalizer::<GreptimeDBMetricNormalize>::default();
+
+        let sink = request
+            .batch_sink(
+                GreptimeDBRetryLogic,
+                greptime_service,
+                MetricsBuffer::new(batch.size),
+                batch.timeout,
+            )
+            .with_flat_map(move |event: Event| {
+                stream::iter({
+                    let byte_size = event.size_of();
+                    let json_size = event.estimated_json_encoded_size_of();
+                    normalizer
+                        .normalize(event.into_metric())
+                        .map(|metric| Ok(EncodedEvent::new(metric, byte_size, json_size)))
+                })
+            })
+            .sink_map_err(|e| error!(message = "Fatal greptimedb sink error.", %e));
+
+        Ok(VectorSink::from_event_sink(sink))
```

Comment:
We are deprecating the use of Event sinks and are moving towards using Event Stream sinks. This will need to be restructured a little to use this.

```suggestion
        Ok(VectorSink::from_event_streamsink(sink))
```

You can find a couple of tutorials [here](https://github.com/vectordotdev/vector/tree/master/docs/tutorials/sinks) that should guide you towards this. Have a look at the [Splunk Hec Metrics sink](https://github.com/vectordotdev/vector/tree/master/src/sinks/splunk_hec/metrics) which I believe is fairly similar to what you are trying to do here.